pam/internal/adapter/gdmmodel.go

```diff
@@ -0,0 +1,277 @@
+			}
+			commands = append(commands, sendEvent(ChangeStage{res.StageChanged.Stage}))
+		}
+	}
+	return tea.Batch(commands...)
+}
+
+func (m *gdmModel) emitEvent(event gdm.Event) tea.Cmd {
+	return func() tea.Msg {
+		return m.emitEventSync(event)
+	}
+}
+
+func (m *gdmModel) emitEventSync(event gdm.Event) tea.Msg {
+	err := gdm.EmitEvent(m.pamMTx, event)
+	log.Debug(context.TODO(), "EventSend", event, "result", err)
+	if err != nil {
+		return pamError{
+			status: pam.ErrSystem,
+			msg:    fmt.Sprintf("Sending GDM event failed: %v", err),
+		}
+	}
+	return nil
+}
+
+func (m gdmModel) Update(msg tea.Msg) (gdmModel, tea.Cmd) {
+	switch msg := msg.(type) {
+	case gdmPollDone:
+		return m, tea.Sequence(
+			tea.Tick(gdmPollFrequency, func(time.Time) tea.Msg { return nil }),
+			m.pollGdm())
+
+	case userSelected:
+		return m, m.emitEvent(&gdm.EventData_UserSelected{
+			UserSelected: &gdm.Events_UserSelected{UserId: msg.username},
+		})
+
+	case brokersListReceived:
+		return m, m.emitEvent(&gdm.EventData_BrokersReceived{
+			BrokersReceived: &gdm.Events_BrokersReceived{BrokersInfos: msg.brokers},
+		})
+
+	case brokerSelected:
+		return m, m.emitEvent(&gdm.EventData_BrokerSelected{
+			BrokerSelected: &gdm.Events_BrokerSelected{BrokerId: msg.brokerID},
+		})
+
+	case authModesReceived:
+		return m, m.emitEvent(&gdm.EventData_AuthModesReceived{
+			AuthModesReceived: &gdm.Events_AuthModesReceived{AuthModes: msg.authModes},
+		})
+
+	case authModeSelected:
+		return m, m.emitEvent(&gdm.EventData_AuthModeSelected{
+			AuthModeSelected: &gdm.Events_AuthModeSelected{AuthModeId: msg.id},
+		})
+
+	case UILayoutReceived:
+		return m, sendEvent(m.emitEventSync(&gdm.EventData_UiLayoutReceived{
+			UiLayoutReceived: &gdm.Events_UiLayoutReceived{UiLayout: msg.layout},
+		}))
+
+	case startAuthentication:
+		if m.waitingAuth {
+			log.Warning(context.TODO(), "Ignored authentication start request while one is still going")
+			return m, nil
+		}
+		m.waitingAuth = true
+		return m, sendEvent(m.emitEventSync(&gdm.EventData_StartAuthentication{
+			StartAuthentication: &gdm.Events_StartAuthentication{},
+		}))
+
+	case isAuthenticatedResultReceived:
+		access := msg.access
+		authMsg, err := dataToMsg(msg.msg)
+		if err != nil {
+			return m, sendEvent(pamError{status: pam.ErrSystem, msg: err.Error()})
+		}
+
+		switch access {
+		case responses.AuthGranted:
+		case responses.AuthDenied:
+		case responses.AuthCancelled:
+			return m, sendEvent(isAuthenticatedCancelled{})
+		case responses.AuthRetry:
+		case responses.AuthNext:
+		default:
+			if authMsg == "" {
+				return m, sendEvent(isAuthenticatedResultReceived{
+					access: responses.AuthDenied,
+					msg:    `{"message": "No valid authentication result"}`,
+				})
+			}
+			return m, sendEvent(isAuthenticatedResultReceived{
+				access: responses.AuthDenied,
+				msg:    fmt.Sprintf(`{"message": "No valid authentication result: %s"}`, authMsg),
+			})
```

Comment:
We don't care about the message if we don't recognize the returned access, so it should be something like:
```suggestion
			return m, sendEvent(isAuthenticatedResultReceived{
				access: responses.AuthDenied,
				msg:    fmt.Sprintf(`{"message": "Access %q is not valid"}`, access),
			})
```